### DISCUSSION TITLE
Component Proposal: Toast

### DISCUSSION THREAD

**Discussion by brandonlenz** [uswds/uswds#5770] [Proposals]
### Component Overview

USWDS currently does not offer a [toast](https://spectrum.adobe.com/page/toast/) (or similar) component for displaying brief messages to users.

Using alerts are not always convenient as they often push content around on the page if shown/hidden dynamically, which makes them intrusive.

### Describe the solution you'd like

Evaluate the need for a Toast, accessibility concerns and best practices, etc, and implement a suitable component

### Describe alternatives you've considered

If there are better evaluated alternatives to achieve similar effect, document and close this issue

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://designsystem.digital.gov/about/community/#community-conduct).
- [X] I checked the [current issues](https://github.com/uswds/uswds/issues?q=is%3Aissue+is%3Aopen+label%3A%22Type%3A+Feature+Request%22) for duplicate feature requests.

**Comment by amycole501** [comment]
This request has received a number of votes (note that it was converted from an issue earlier this year).

We'd like to know more about the use cases for a toast component. Is the request coming from a place of needing an option to an alert or error message? Do the toast alerts need to disappear automatically or be dismissed by the user? Live examples would be great so we can best follow up on this request. We also want as many contributions as possible by a wide range of teams. Knowing more will allow us to provide an informed recommendation.

**Reply by jbouder** [reply]
There is definitely a need from my end. To me the main differences between an Alert and a Toast are:
- A toast appears temporarily, an Alert usually persists
- A toast "floats" above the UI, where as an Alert is usually inline

In terms of specs, I like the UX of the AntD Notifications: https://3x.ant.design/components/notification/. 

I would expect the following from a Toast:
- Should appear temporarily by default, but should be configurable
- Should float above the UI
- Should appear in the top right by default, but should be configurable
- Should allow different types of Toasts, aligning with USWDS alerts types makes sense to me
- Should stack Toasts if more than 1 is displayed at a time

**Comment by annepetersen** [comment]
Thanks for your proposal, @brandonlenz, and your reply here, @jbouder.

## Toast proposal decision
The USWDS core team reviewed this, and we’ve decided **not to move forward** with developing this toast component (or, as originally proposed, [feature](https://github.com/uswds/uswds/issues/5071)). 

### Background
This proposal got 11 votes by the time of this response, which I wanted to note since folks can continue to add/subtract their upvotes later. 

We requested more use cases here, and directed our community to this discussion in our monthly call and newsletter multiple times, but we’ve only received the one reply. We did do our own evaluation too, though.

### Reasons
We based this decision primarily on **accessibility concerns** with the vast majority of implementations we looked at. These include:

- **No native HTML element**: no built-in HTML element for a toast exists ([see this examination of toast)](https://adrianroselli.com/2020/01/defining-toast-messages.html)
- [**Timing concerns** (WCAG 2.2.1)](https://www.w3.org/WAI/WCAG22/Understanding/timing-adjustable.html): toasts that disappear automatically probably won’t meet the requirement to allow timing adjustments, which means people could miss important information
- [**Keyboard navigation issues** (WCAG 2.1.2](https://www.w3.org/WAI/WCAG21/Understanding/no-keyboard-trap.html)): since a toast overlays the user interface, it can cause problems with keyboard focus — some implementations allow accidental focus on elements behind the toast
- [**Dismissibility** (WCAG 1.4.13)](https://www.w3.org/WAI/WCAG21/Understanding/content-on-hover-or-focus.html): messages have to be dismissible by the user, and not disappear automatically, since again, people could miss critical info
- [**Focus order and sequence**](https://www.w3.org/WAI/WCAG21/Understanding/focus-order.html): if the toast appears at the end of the DOM (document structure), it can create issues with focus and the logical sequence of content
- [**Zoom magnification visibility issues**](https://www.w3.org/WAI/WCAG21/Understanding/reflow.html): people using large screens, or zooming in (200% or more) might not see it
- **Reliance on visual cues**: many toast components use motion or visual effects, which may not be perceivable or understandable for everyone, or problematic in other ways
- **Screen reader issues**: screen readers might read multiple toast alerts at once, interrupting the page flow — and since toasts often appear at the end of the DOM, they can often be missed 
- **Modals**: toasts and modals may not always function together well
- [**Focus visibility** (WCAG 2.4.11)](https://www.w3.org/WAI/WCAG22/Understanding/focus-not-obscured-minimum): depending on how the toast is displayed, it can obscure focus, violating this guideline

We understand that many components can come with accessibility challenges, but from our research and experience, we believe the toast component **introduces more complexity than it solves**. So in the end, toast’s **usability and accessibility risks outweigh its potential benefits**.

Usability testing can help you determine if a toast is truly necessary, but we believe using proper page structure and semantic HTML will be a better way to both meet accessibility guidelines _and_ make sure important messages aren’t missed.

### Other options

When reaching for a toast-like component, you could ask yourself, "What message do I need to communicate with this?” Can it be permanent or dismissable, part of the DOM, and integrated into the user interface instead?

And instead of creating your own separate toast component, you could consider [proposing a variation](https://designsystem.digital.gov/about/contribute/#propose-a-feature-request-or-enhancement-2) of the [alert component](https://designsystem.digital.gov/components/alert/) that’s dismissible, maybe positioned at the top of the screen. This way, everyone — regardless of device or assistive technology (or, potentially, neurotype) — can easily understand and interact with it.

### Why
We realize this decision is probably disappointing, but hope you understand that we **must** prioritize accessibility in USWDS. Both our [design principles](https://designsystem.digital.gov/design-principles/#embrace-accessibility-2) — which all teams using USWDS should be using — and our [product values](https://designsystem.digital.gov/design-principles/#embrace-accessibility-2) note how fundamental accessibility is to serving government’s various audiences, how we need to build these considerations into every decision we make, and how we need to go beyond basic compliance.

Let us know if you have any questions. Thanks!

**Reply by jbouder** [reply]
And thank you for the due diligence!
